Dialog/CreditEditor_TableView.cpp

```diff
@@ -0,0 +1,237 @@
+#include <QMenu>
+#include <iostream>
+#include <QPainter>
+#include <QKeyEvent>
+
+
+CreditEditor_TableView::CreditEditor_TableView(QWidget *parent) :
+    QTableView(parent)
+{
+    menu=new QMenu("Menu", this);
+    action_one_tile = new QAction("One Tile",menu);
+    action_upper_tile = new QAction("Upper Tile",menu);
+    action_lower_tile = new QAction("Lower Tile",menu);
+    action_delete = new QAction("Delete",menu);
+
+    menu->addAction(action_one_tile);
+    menu->addAction(action_upper_tile);
+    menu->addAction(action_lower_tile);
+    menu->addAction(action_delete);
+
+    this->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(this, SIGNAL(customContextMenuRequested(const QPoint &)), this, SLOT(showContextMenu(const QPoint &)));
+    connect(menu, SIGNAL(triggered(QAction *)), this, SLOT(doAction(QAction *)));
+}
+
+
+
+/// <summary>
+/// Called each time a cell is changed
+/// </summary>
+/// <param name="topLeft">
+/// The top left QModelIndex
+/// <param name="bottomRight">
+/// Unused
+/// <param name="roles">
+/// Unused
+/// </param>
+void CreditEditor_TableView::dataChanged(const QModelIndex &topLeft, const QModelIndex &bottomRight, const QVector<int> &roles) {
+
+    QString first_letter=topLeft.data(Qt::DisplayRole).value<QString>();
+    QString role=topLeft.data(Qt::BackgroundRole).value<QColor>().name();
+
+    std::string string = first_letter.toUpper().toUtf8().constData();
+
+    if (string.size() > 0) {
+        std::string role_string = role.toUtf8().constData();
+        std::string first_letter_string(1,string.at(0));
+
+        //If the tile is not valid
+        if (CreditsEditDialog::CreditTileReverseMap[std::make_pair(first_letter_string,role_string)] == 0) {
+
+            //If the tile is still not valid with the default background we delete it
+            if (CreditsEditDialog::CreditTileReverseMap[std::make_pair(first_letter_string,"#8080ff")] == 0) {
+                deleteFunction(topLeft);
+            }
+        }
+    }
+
+}
+
+/// <summary>
+/// Show a popup menu when a right click is made on the QTableView.
+/// </summary>
+/// <param name="pos">
+/// The position of cursor.
+/// </param>
+void CreditEditor_TableView::showContextMenu(const QPoint &pos) {
+    menu->exec(this->mapToGlobal(pos),action_one_tile);
+}
+
+
+/// <summary>
+/// When the popup menu is clicked set selected cells in the described category
+/// </summary>
+/// <param name="action">
+/// Action from the popup menu (One Tile, Upper Tile, Lower Tile, Delete).
+/// </param>
+void CreditEditor_TableView::doAction(QAction * action) {
+     QString action_text=action->text();
+
+     QModelIndexList index_list=this->selectionModel()->selectedIndexes();
+     for (int i=0;i<index_list.size();i++) {
+         QModelIndex model_index=index_list.at(i);
+         int column=model_index.column();
+
+         //If we are in the editable zone
+         if (column < 30) {
+             if (action_text.compare("One Tile") == 0) {
```

Comment:
Is there some way we can check the action index or make this determination without a string comparison? This will make it difficult to provide multilingual translations in the future via i18n

Reply:
This has not been addressed yet

Reply:
Replaced by
example : if (action == actionOneTile).
for each
(Cannot use UI because I couldn't setup a grid layout via the editor (and putting 13 tabs was easier programmatically).
I needed a grid layout with only 1 row and column to make it expandable in all direction)

Reply:
This is an acceptable fix.